model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py

```diff
@@ -0,0 +1,257 @@
+import math
+import pathlib
+from typing import AsyncIterator, Type
+
+import numpy as np
+
+from dffml import (
+    Accuracy,
+    Feature,
+    Features,
+    ModelNotTrained,
+    Record,
+    SimpleModel,
+    SourcesContext,
+    config,
+    entrypoint,
+    field,
+)
+
+
+def split(a_list, k):
+    L = int(k * len(a_list))
+    return a_list[:L], a_list[L:]
+
+
+def estimateGaussian(X):
+
+    n = np.size(X, 1)
+    m = np.size(X, 0)
+    mu = np.zeros((n, 1))
+    sigma2 = np.zeros((n, 1))
+
+    mu = np.reshape((1 / m) * np.sum(X, 0), (1, n))
+    sigma2 = np.reshape(((1 / m) * np.sum(np.square(X - mu), 0)), (1, n))
+    return mu, sigma2
+
+
+def multivariateGaussian(X, mu, sigma2):
+    n = np.size(sigma2, 1)
+    m = np.size(sigma2, 0)
+
+    if n == 1 or m == 1:
+        sigma2 = np.diag(sigma2[0, :])
+
+    X = X - mu
+    pi = math.pi
+    det = np.linalg.det(sigma2)
+    inv = np.linalg.inv(sigma2)
+    val = np.reshape(
+        (-0.5) * np.sum(np.multiply((X @ inv), X), 1), (np.size(X, 0), 1)
+    )
+
+    p = np.power(2 * pi, n / -2) * np.power(det, -0.5) * np.exp(val)
+
+    return p
+
+
+def findIndices(binVec):
+    l = []
+    for i in range(len(binVec)):
+        if binVec[i] == 1:
+            l.append(i)
+    return l
+
+
+def selectThreshold(Yval, pval):
+    bestEpsilon = 0
+    bestF1 = 0
+    F1 = 0
+    stepsize = (np.max(pval) - np.min(pval)) / 1000
+    epsVec = np.arange(np.min(pval), np.max(pval), stepsize)
+    lenv = len(epsVec)
+
+    for eps in range(lenv):
+        epsilon = epsVec[eps]
+        pred = (pval < epsilon).astype(int)
+        F1 = getF1(Yval, pred)
+        if F1 > bestF1:
+            bestF1 = F1
+            bestEpsilon = epsilon
+
+    return bestF1, bestEpsilon
+
+
+def getF1(Yval, pred):
+    prec, rec = 0, 0
+    tp, fp, fn = 0, 0, 0
+
+    try:
+        fp = np.sum(((pred == 1) & (Yval == 0)).astype(int))
+        tp = np.sum(((pred == 1) & (Yval == 1)).astype(int))
+        fn = np.sum(((pred == 0) & (Yval == 1)).astype(int))
+
+        prec = tp / (tp + fp)
+        rec = tp / (tp + fn)
+        F1 = 2 * prec * rec / (prec + rec)
+    except ZeroDivisionError:
+        raise
+    return F1
+
+
+@config
+class AnomalyModelConfig:
+    features: Features = field("Features to train on")
+    predict: Feature = field("Label or the value to be predicted")
+    directory: pathlib.Path = field("Directory where state should be saved")
+    k: float = field("Validation set size", default=0.8)
+
+
+@entrypoint("anomalydetection")
+class AnomalyModel(SimpleModel):
```

Comment:
https://intel.github.io/dffml/master/contributing/consoletest.html

For writing out a file:


```rst
.. code-block::
    :test:
    :filepath: training.csv

    x,y
    0,1
```